# Préprocess au NLP

Tout d’abord, install le paquet NLTK avec le gestionnaire de paquets + pip +:

In [1]:
#pip install nltk==3.3

Importation des librairies

In [25]:
import nltk
import pandas as pd
import numpy as np
from nltk.corpus import twitter_samples
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ouizb\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

Téléchargez des commentaires Tripadvisor

In [26]:
df = pd.read_csv('resampled_comments_2.csv')
df

Unnamed: 0                                            comment  rating  \
0               1  Accueil sympa pas mal d’attente au final le ke...     3.0   
1               7  L’accueil est bonne mais les sandwichs mauvais...     1.0   
2               1  En balade dans le quartier nous cherchons un k...     3.0   
3               6  kebab identique à ailleurs cela reste un kebab...     3.0   
4               5  Au moment où je rédige cet avis le restaurant ...     1.0   
...           ...                                                ...     ...   
13305           5  Ce n'est pas difficile face à l'Elysee vous re...     5.0   
13306           7  Nous avons vraiment passé un excellent moment ...     5.0   
13307           3  Nous arrivons au service de 21h heure à laquel...     5.0   
13308           1  Ici comme à l'Eden Roc ou dans un autre hotel ...     5.0   
13309           2  Excellente pizzaria. Un accueil très chaleureu...     5.0   

                                              bonus_info   city sentiment  
0      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
1      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris  Negative  
2      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
3      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
4      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris  Negative  
...                                                  ...    ...       ...  
13305  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13306  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13307  Détails FOURCHETTE DE PRIX 2 628 AMD - 13 140 ...  Paris  Positive  
13308  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13309  Emplacement et coordonnées 43 Rue de Charenton...  Paris  Positive  

[13310 rows x 6 columns]

### On  a séparer les commentaires en fonction de leur note en 3 catégories:
- supérieur ou égale à 4 => positif
- égale à 3 => neutre
- inférieur ou égale à 2 => négatif

Je prends juste les commentaires, les notes  et les sentiments pour le moment.

In [27]:
comments = df[['comment', 'rating', 'sentiment']]
comments.head(10)

comment  rating sentiment
0  Accueil sympa pas mal d’attente au final le ke...     3.0   Neutral
1  L’accueil est bonne mais les sandwichs mauvais...     1.0  Negative
2  En balade dans le quartier nous cherchons un k...     3.0   Neutral
3  kebab identique à ailleurs cela reste un kebab...     3.0   Neutral
4  Au moment où je rédige cet avis le restaurant ...     1.0  Negative
5  Comment reconnaître des avis laissés par un pa...     1.0  Negative
6  Ce n’est pas possible que cet endroit soit cla...     1.0  Negative
7  Très surpris du nombre de profil avec 1 seul c...     1.0  Negative
8  Pas de salle juste deux tabourets a la limite ...     1.0  Negative
9  Je suis allée dans ce kebab après avoir vu les...     1.0  Negative

## Suppression du bruit

On met tout le texte en minuscule avnat la tokenization

In [28]:
comments['comment'] = comments['comment'].map(lambda x: x.lower())
comments

C:\Users\ouizb\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


comment  rating sentiment
0      accueil sympa pas mal d’attente au final le ke...     3.0   Neutral
1      l’accueil est bonne mais les sandwichs mauvais...     1.0  Negative
2      en balade dans le quartier nous cherchons un k...     3.0   Neutral
3      kebab identique à ailleurs cela reste un kebab...     3.0   Neutral
4      au moment où je rédige cet avis le restaurant ...     1.0  Negative
...                                                  ...     ...       ...
13305  ce n'est pas difficile face à l'elysee vous re...     5.0  Positive
13306  nous avons vraiment passé un excellent moment ...     5.0  Positive
13307  nous arrivons au service de 21h heure à laquel...     5.0  Positive
13308  ici comme à l'eden roc ou dans un autre hotel ...     5.0  Positive
13309  excellente pizzaria. un accueil très chaleureu...     5.0  Positive

[13310 rows x 3 columns]

## Tokenisation

Tokenization des phrases en mot.

In [29]:
from nltk import RegexpTokenizer
toknizer = RegexpTokenizer(r'''\w'|\w+|[^\w\s]''')
token = comments.apply(lambda row: toknizer.tokenize(row['comment']),axis= 1)
token

0        [accueil, sympa, pas, mal, d, ’, attente, au, ...
1        [l, ’, accueil, est, bonne, mais, les, sandwic...
2        [en, balade, dans, le, quartier, nous, chercho...
3        [kebab, identique, à, ailleurs, cela, reste, u...
4        [au, moment, où, je, rédige, cet, avis, le, re...
                               ...                        
13305    [ce, n', est, pas, difficile, face, à, l', ely...
13306    [nous, avons, vraiment, passé, un, excellent, ...
13307    [nous, arrivons, au, service, de, 21h, heure, ...
13308    [ici, comme, à, l', eden, roc, ou, dans, un, a...
13309    [excellente, pizzaria, ., un, accueil, très, c...
Length: 13310, dtype: object

Suppression de la ponctuation

In [30]:
import string
import re 
#string.punctuation
token_1 = token.apply(lambda x: [item.strip(string.punctuation) for item in x])
token_2 = token_1.apply(lambda x: [re.sub(r'\’', '', item) for item in x])
token_2

0        [accueil, sympa, pas, mal, d, , attente, au, f...
1        [l, , accueil, est, bonne, mais, les, sandwich...
2        [en, balade, dans, le, quartier, nous, chercho...
3        [kebab, identique, à, ailleurs, cela, reste, u...
4        [au, moment, où, je, rédige, cet, avis, le, re...
                               ...                        
13305    [ce, n, est, pas, difficile, face, à, l, elyse...
13306    [nous, avons, vraiment, passé, un, excellent, ...
13307    [nous, arrivons, au, service, de, 21h, heure, ...
13308    [ici, comme, à, l, eden, roc, ou, dans, un, au...
13309    [excellente, pizzaria, , un, accueil, très, ch...
Length: 13310, dtype: object

suppression des nombres

In [31]:
token_4 = token_2.apply(lambda x: [re.sub(r'\d', '', item) for item in x])

print(token_4) 

0        [accueil, sympa, pas, mal, d, , attente, au, f...
1        [l, , accueil, est, bonne, mais, les, sandwich...
2        [en, balade, dans, le, quartier, nous, chercho...
3        [kebab, identique, à, ailleurs, cela, reste, u...
4        [au, moment, où, je, rédige, cet, avis, le, re...
                               ...                        
13305    [ce, n, est, pas, difficile, face, à, l, elyse...
13306    [nous, avons, vraiment, passé, un, excellent, ...
13307    [nous, arrivons, au, service, de, h, heure, à,...
13308    [ici, comme, à, l, eden, roc, ou, dans, un, au...
13309    [excellente, pizzaria, , un, accueil, très, ch...
Length: 13310, dtype: object


## Normalisation

### Stopword Removal : Suppression des mots vides

In [32]:
from nltk.corpus import stopwords 
import nltk 
nltk.download('stopwords')
stop_words = set(stopwords.words('french')) 
stop_token = token_4.apply(lambda x: [item for item in x if item not in stop_words])

print(stop_token)

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ouizb\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
0        [accueil, sympa, mal, , attente, final, kebab,...
1        [, accueil, bonne, sandwichs, mauvais, plus, t...
2        [balade, quartier, cherchons, kebab, voici, tr...
3        [kebab, identique, ailleurs, cela, reste, keba...
4        [moment, où, rédige, cet, avis, restaurant, cl...
                               ...                        
13305    [difficile, face, elysee, remontez, rue, côté,...
13306    [vraiment, passé, excellent, moment, nouveau, ...
13307    [arrivons, service, h, heure, laquelle, réserv...
13308    [ici, comme, eden, roc, autre, hotel, groupe, ...
13309    [excellente, pizzaria, , accueil, très, chaleu...
Length: 13310, dtype: object


### Stemming

In [33]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer(language='french')
stemm = stop_token.apply(lambda x: [stemmer.stem(y) for y in x])

print(stemm)

0        [accueil, symp, mal, , attent, final, kebab, c...
1        [, accueil, bon, sandwich, mauv, plus, tromp, ...
2        [balad, quarti, cherchon, kebab, voic, tripadv...
3        [kebab, ident, ailleur, cel, rest, kebab, tran...
4        [moment, où, rédig, cet, avis, restaur, class,...
                               ...                        
13305    [difficil, fac, elyse, remont, ru, côt, bristo...
13306    [vrai, pass, excellent, moment, nouveau, resta...
13307    [arrivon, servic, h, heur, laquel, réserv, dîn...
13308    [ici, comm, eden, roc, autr, hotel, group, lev...
13309    [excellent, pizzari, , accueil, tres, chaleur,...
Length: 13310, dtype: object


### Lemmatization

In [11]:
#pip install git+https://github.com/ClaudeCoulombe/FrenchLefffLemmatizer.git

In [34]:
from french_lefff_lemmatizer.french_lefff_lemmatizer import FrenchLefffLemmatizer
lemmatizer = FrenchLefffLemmatizer()
lemm = stemm.apply(lambda x: [lemmatizer.lemmatize(y) for y in x])
lemm

0        [accueil, symp, mal, , attent, final, kebab, c...
1        [, accueil, bon, sandwich, mauv, plus, tromp, ...
2        [balad, quarti, cherchon, kebab, voic, tripadv...
3        [kebab, ident, ailleur, cel, rest, kebab, tran...
4        [moment, où, rédig, cet, avis, restaur, class,...
                               ...                        
13305    [difficil, fac, elyse, remont, ru, côt, bristo...
13306    [vrai, pass, excellent, moment, nouveau, resta...
13307    [arrivon, servic, heure, heur, laquel, réserv,...
13308    [ici, comm, eden, roc, autr, hotel, group, lev...
13309    [excellent, pizzari, , accueil, tres, chaleur,...
Length: 13310, dtype: object

Maintenant, assemblons les tokens traités ensemble.

In [35]:
for i in range(len(lemm)):
    lemm[i] = ' '.join(lemm[i])


In [36]:
# On rajoute à notre dataframe les données lemmatiser comme nouvelle colonne
comments['lemmatiser_com'] = lemm
comments

C:\Users\ouizb\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


comment  rating sentiment  \
0      accueil sympa pas mal d’attente au final le ke...     3.0   Neutral   
1      l’accueil est bonne mais les sandwichs mauvais...     1.0  Negative   
2      en balade dans le quartier nous cherchons un k...     3.0   Neutral   
3      kebab identique à ailleurs cela reste un kebab...     3.0   Neutral   
4      au moment où je rédige cet avis le restaurant ...     1.0  Negative   
...                                                  ...     ...       ...   
13305  ce n'est pas difficile face à l'elysee vous re...     5.0  Positive   
13306  nous avons vraiment passé un excellent moment ...     5.0  Positive   
13307  nous arrivons au service de 21h heure à laquel...     5.0  Positive   
13308  ici comme à l'eden roc ou dans un autre hotel ...     5.0  Positive   
13309  excellente pizzaria. un accueil très chaleureu...     5.0  Positive   

                                          lemmatiser_com  
0      accueil symp mal  attent final kebab correct s...  
1       accueil bon sandwich mauv plus tromp command ...  
2      balad quarti cherchon kebab voic tripadvisor i...  
3      kebab ident ailleur cel rest kebab transcend  ...  
4      moment où rédig cet avis restaur class eme tri...  
...                                                  ...  
13305  difficil fac elyse remont ru côt bristol  seco...  
13306  vrai pass excellent moment nouveau restaur esp...  
13307  arrivon servic heure heur laquel réserv dîn  a...  
13308  ici comm eden roc autr hotel group levur chimi...  
13309  excellent pizzari  accueil tres chaleur pizz v...  

[13310 rows x 4 columns]

###  Fonctionnalités TF-IDF
Cette méthode prend en compte non seulement l'occurrence d'un mot dans un seul commentaire (ou tweet) mais aussi dans le corpus entier.
TF-IDF fonctionne en pénalisant les mots communs en leur attribuant des poids inférieurs tout en donnant de l'importance aux mots qui sont rares dans l'ensemble du corpus mais qui apparaissent en bon nombre dans peu de commentaires.

Termes importants liés à TF-IDF:
- TF = (Nombre de fois que le terme t apparaît dans un document) / (Nombre de termes dans le document)
- IDF = log (N / n), où, N est le nombre de documents et n est le nombre de documents dans lesquels un terme t est apparu.
- TF-IDF = TF * IDF

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(max_df = 0.90 , min_df = 2 , max_features = 1000 , stop_words = stop_words, ngram_range=(1, 2)) 
# TF-IDF matrice de fonctionnalités
X = tfidf_vectorizer.fit_transform(comments['lemmatiser_com'])


## Construction de modèles: analyse des sentiments

In [16]:
y = comments['sentiment']

#### Split trani et test set

In [17]:
from sklearn.model_selection import train_test_split
# TF-IDF matrice
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

### Random Forest

In [18]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

# TF-IDF matrice
rf = RandomForestClassifier()
rf.fit(X_train, y_train)
scores = cross_val_score(rf, X_train, y_train, cv=5)
print(scores, "Avg accuracy Random Forest avec TF-IDF matrice: {}".format(np.mean(scores)))

[0.79624413 0.81314554 0.80985915 0.79051198 0.79802724] Avg accuracy Random Forest avec TF-IDF matrice: 0.8015576093164591


### Naive bayes

####  Naive bayes multinomial

In [19]:
from sklearn.naive_bayes import MultinomialNB
# TF-IDF matrice
mnb = MultinomialNB()
mnb.fit(X_train.todense(), y_train)
scores_mnb = cross_val_score(mnb, X_train, y_train, cv=5)
print(scores_mnb, "Avg accuracy MultinomialNB avec TF-IDF matrice: {}".format(np.mean(scores_mnb)))

[0.70187793 0.70938967 0.71126761 0.70596524 0.7012682 ] Avg accuracy MultinomialNB avec TF-IDF matrice: 0.7059537308397118


### SVM

In [20]:
from sklearn.svm import SVC
# TF-IDF matrice
svc_model = SVC(gamma='auto')
svc_model.fit(X_train, y_train)
scores_svm = cross_val_score(svc_model, X_train, y_train, cv=5)
print(scores_svm, "Avg accuracy SVM avec TF-IDF matrice: {}".format(np.mean(scores_svm)))

[0.38920188 0.38920188 0.38873239 0.38891498 0.38938469] Avg accuracy SVM avec TF-IDF matrice: 0.38908716428837625


Random Forest à l'accuracy la plus élevée avec TF-IDF matrice avec 80.16%.

Random Forest à l'accuracy la plus élevée avec bag-of-words matrice avec 80.24%.

## Ensemble learning

### Stacking

In [21]:
from sklearn.ensemble import StackingClassifier

estimators = [('Random forest', rf), ("Naive bayes", mnb),("SVM", svc_model)]
reg = StackingClassifier(estimators=estimators, final_estimator=SVC())
reg.fit(X_train, y_train)
scores_reg = reg.score(X_train, y_train)

print(scores_reg, "Avg accuracy RandomForestClassifier: {}".format(np.mean(scores_reg)))
y_pred_reg = reg.predict(X_test)
print(accuracy_score(y_test, y_pred_reg))

0.9930503380916604 Avg accuracy RandomForestClassifier: 0.9930503380916604
0.8441021788129226


### Bagging

In [22]:
from sklearn.ensemble import BaggingClassifier
from sklearn import tree
from sklearn.metrics import accuracy_score, classification_report, precision_score, recall_score 
from sklearn.metrics import confusion_matrix, precision_recall_curve, roc_curve, auc, log_loss
from sklearn.metrics import accuracy_score
model_bag = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))
model_bag.fit(X_train, y_train)
model_bag.score(X_test,y_test)

y_pred_bag = model_bag.predict(X_test)


print(accuracy_score(y_test, y_pred_bag))


0.7637114951164538


### Ada boost

In [23]:
from sklearn.ensemble import AdaBoostClassifier

model_ada = AdaBoostClassifier(random_state=1)
model_ada.fit(X_train, y_train)
#model.score(x_test,y_test)

y_pred_ada = model_ada.predict(X_test)

print(accuracy_score(y_test, y_pred_ada))

0.6036814425244177


### Voting

In [24]:
from sklearn.ensemble import VotingClassifier
voting = VotingClassifier(estimators=[('SVM', svc_model), ('Naive bayes', mnb),('Random forest', rf) ], voting='hard')
voting.fit(X_train, y_train)
scores_voting = cross_val_score(voting, X_train, y_train, cv=5)
print(scores_voting, "Avg accuracy Voting: {}".format(np.mean(scores_voting)))
y_pred_voting = voting.predict(X_test)
print(accuracy_score(y_test, y_pred_voting))

[0.7314554  0.74507042 0.75164319 0.73320808 0.74025364] Avg accuracy Voting: 0.7403261466402927
0.7453042824943651
